tests/pipelines/test_cli.py

```diff
@@ -124,6 +124,19 @@ def test_delete_cli_incorrect_parameters(pipelines_api_mock, tmpdir):
     assert pipelines_api_mock.delete.call_count == 0
 
 
+@provide_conf
+def test_deploy_spec_updated_with_id_if_pipeline_id_not_in_spec(pipelines_api_mock, tmpdir):  # noqa
+    path = tmpdir.join('/spec.json').strpath
+    spec_with_no_id = '{"name": "no id"}'
+    with open(path, 'w') as f:
+        f.write(spec_with_no_id)
+    runner = CliRunner()
+    runner.invoke(cli.deploy_cli, ['--spec', path])
+    with open(path, 'r') as f:
+        updated_spec = json.loads(f.read())
+    assert 'id' in updated_spec
```

Comment:
Couple missing tests for not touching the ID if it's valid, and what happens if the id is an empty string.

Reply:
Let's also test for invalid json?

Reply:
Invalid json input is tested in `_read_spec`. I think validate should raise an error for an empty or invalid id and I'll add a test for the case when an id is provided.